gRPC codegen fails for io.envoyproxy.controlplane:api

### Describe the bug

I stumbled upon this problem while testing envoy control plane api.

gRPC codegen fails with an exception when adding `quarkus.generate-code.grpc.scan-for-proto=io.envoyproxy.controlplane:api` to scan envoy controlplane api for proto files.

```
Caused by: io.quarkus.bootstrap.prebuild.CodeGenException: Failed to generate java files from proto file in C:\test\code-with-quarkus\build\classes\java\quarkus-generated-sources\proto\.
        at io.quarkus.grpc.deployment.GrpcCodeGen.trigger(GrpcCodeGen.java:177)
        at io.quarkus.deployment.CodeGenerator.lambda$trigger$4(CodeGenerator.java:204)
        at io.quarkus.deployment.CodeGenerator.callWithClassloader(CodeGenerator.java:179)
        at io.quarkus.deployment.CodeGenerator.trigger(CodeGenerator.java:201)
        at io.quarkus.deployment.CodeGenerator.initAndRun(CodeGenerator.java:78)
        ... 28 more
Caused by: java.io.IOException: Cannot run program "C:\test\code-with-quarkus\build\com.google.protobuf-protoc-windows-x86_64-exe": CreateProcess error=206, The filename or extension is too long
        at io.quarkus.deployment.util.ProcessUtil.launchProcess(ProcessUtil.java:38)
        at io.quarkus.grpc.deployment.GrpcCodeGen.trigger(GrpcCodeGen.java:165)
        ... 32 more
Caused by: java.io.IOException: CreateProcess error=206, The filename or extension is too long
```

Note that the io.envoyproxy.controlplane:api artifact also contains generated code, so it is possible to use the code from the artifact.

### Expected behavior

No exception occurs

### Actual behavior

`java.io.IOException: CreateProcess error=206, The filename or extension is too long` occurs

### How to Reproduce?

Reproducer: https://github.com/ls-bit/quarkus-grpc-codegen-issue

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.22631.3007]

### Output of `java -version`

openjdk version "17.0.8" 2023-07-18 OpenJDK Runtime Environment GraalVM CE 17.0.8+7.1 (build 17.0.8+7-jvmci-23.0-b15) OpenJDK 64-Bit Server VM GraalVM CE 17.0.8+7.1 (build 17.0.8+7-jvmci-23.0-b15, mixed mode, sharing)

### Quarkus version or git rev

Quarkus 3.7.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.5

### Additional information

_No response_

--- a/docs/src/main/asciidoc/grpc-generation-reference.adoc
+++ b/docs/src/main/asciidoc/grpc-generation-reference.adoc
@@ -288,3 +288,12 @@ plugins {
 
 IMPORTANT: It is recommended to package the `proto` files in a dependency instead of the generated classes, so Quarkus can generate optimized classes.
 Refer to the <<scan-for-proto, dedicated section>> for more information.
+
+== Argument files
+
+When the `protoc` command line exceeds the maximum command length, you can ask Quarkus to use an argument file to pass the arguments to the `protoc` command.
+
+To enable this feature, set the `quarkus.generate-code.grpc.use-arg-file` property in your `application.properties` file to `true`.
+
+If you are on Windows, and the command line exceeds 8190 characters, Quarkus automatically uses an argument file to pass the arguments to the `protoc` command.
+

--- a/extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java
+++ b/extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java
@@ -5,7 +5,10 @@ import static java.lang.Boolean.TRUE;
 import static java.util.Arrays.asList;
 
 import java.io.BufferedWriter;
+import java.io.File;
 import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -61,6 +64,8 @@ public class GrpcCodeGen implements CodeGenProvider {
     private static final String DESCRIPTOR_SET_OUTPUT_DIR = "quarkus.generate-code.grpc.descriptor-set.output-dir";
     private static final String DESCRIPTOR_SET_FILENAME = "quarkus.generate-code.grpc.descriptor-set.name";
 
+    private static final String USE_ARG_FILE = "quarkus.generate-code.grpc.use-arg-file";
+
     private Executables executables;
     private String input;
 
@@ -103,6 +108,9 @@ public class GrpcCodeGen implements CodeGenProvider {
         Path workDir = context.workDir();
         Path inputDir = CodeGenProvider.resolve(context.inputDir());
         Set<String> protoDirs = new LinkedHashSet<>();
+
+        boolean useArgFile = context.config().getOptionalValue(USE_ARG_FILE, Boolean.class).orElse(false);
+
         try {
             List<String> protoFiles = new ArrayList<>();
             if (Files.isDirectory(inputDir)) {
@@ -160,6 +168,21 @@ public class GrpcCodeGen implements CodeGenProvider {
 
                 command.addAll(protoFiles);
 
+                // Estimate command length to avoid command line too long error
+                int commandLength = command.stream().mapToInt(String::length).sum();
+                // 8191 is the maximum command line length for Windows
+                if (useArgFile || (commandLength > 8190 && OS.determineOS() == OS.WINDOWS)) {
+                    File argFile = File.createTempFile("grpc-protoc-params", ".txt");
+                    argFile.deleteOnExit();
+
+                    try (PrintWriter writer = new PrintWriter(argFile, StandardCharsets.UTF_8)) {
+                        for (int i = 1; i < command.size(); i++) {
+                            writer.println(command.get(i));
+                        }
+                    }
+
+                    command = new ArrayList<>(Arrays.asList(command.get(0), "@" + argFile.getAbsolutePath()));
+                }
                 ProcessBuilder processBuilder = new ProcessBuilder(command);
 
                 final Process process = ProcessUtil.launchProcess(processBuilder, context.shouldRedirectIO());
@@ -170,6 +193,7 @@ public class GrpcCodeGen implements CodeGenProvider {
                 }
                 postprocessing(context, outDir);
                 log.info("Successfully finished generating and post-processing sources from proto files");
+
                 return true;
             }
         } catch (IOException | InterruptedException e) {

--- a/integration-tests/grpc-plain-text-gzip/src/main/resources/application.properties
+++ b/integration-tests/grpc-plain-text-gzip/src/main/resources/application.properties
@@ -12,3 +12,6 @@ quarkus.http.enable-compression=true
 %vertx.quarkus.grpc.clients.hello.port=8081
 %vertx.quarkus.grpc.clients.hello.use-quarkus-grpc-client=true
 %vertx.quarkus.grpc.server.use-separate-server=false
+
+# Force the usage of an arg file
+quarkus.generate-code.grpc.use-arg-file=true

